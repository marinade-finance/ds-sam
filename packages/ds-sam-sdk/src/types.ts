import Decimal from 'decimal.js'

export type AuctionResult = {
  auctionData: AuctionData
  winningTotalPmpe: number
}

export type AuctionData = Omit<AggregatedData, 'validators'> & {
  validators: AuctionValidator[]
}

export type StakeAmounts = {
  networkTotalSol: number
  marinadeMndeTvlSol: number
  marinadeSamTvlSol: number
  marinadeRemainingMndeSol: number
  marinadeRemainingSamSol: number
}

export type AggregatedData = {
  validators: AggregatedValidator[]
  rewards: Rewards
  stakeAmounts: StakeAmounts
  blacklist: Set<string>
}

export type EpochStats = {
  epoch: number
  totalActivatedStake: Decimal
  marinadeActivatedStake: Decimal
  voteCredits: number
}

export type ValidatorAuctionStake = {
  externalActivatedSol: number
  marinadeMndeTargetSol: number
  marinadeSamTargetSol: number
}

export type AuctionValidator = AggregatedValidator & {
  revShare: RevShare
  mndeEligible: boolean
  samEligible: boolean
  auctionStake: ValidatorAuctionStake
  lastCapConstraint: AuctionConstraint | null
  stakePriority: number
  unstakePriority: number
}

export type AggregatedValidator = {
  voteAccount: string
  clientVersion: string
  voteCredits: number
  aso: string
  country: string
  bondBalanceSol: number | null
  totalActivatedStakeSol: number
  marinadeActivatedStakeSol: number
  inflationCommissionDec: number
  mevCommissionDec: number | null
  bidCpmpe: number | null
  maxStakeWanted: number | null
  mndeStakeCapIncrease: number
  mndeVotesSolValue: number
  epochStats: EpochStats[]
}

export type Rewards = {
  inflationPmpe: number
  mevPmpe: number
}

export type RevShare = {
  totalPmpe: number
  inflationPmpe: number
  mevPmpe: number
  bidPmpe: number
  auctionEffectiveBidPmpe: number
}

export type AuctionConstraintsConfig = {
  totalCountryStakeCapSol: number
  totalAsoStakeCapSol: number
  marinadeCountryStakeCapSol: number
  marinadeAsoStakeCapSol: number
  marinadeValidatorStakeCapSol: number
}

export enum AuctionConstraintType {
  COUNTRY = 'COUNTRY',
  ASO = 'ASO',
  VALIDATOR = 'VALIDATOR',
  BOND = 'BOND',
  MNDE = 'MNDE',
}

export type AuctionConstraint = {
  constraintType: AuctionConstraintType
  constraintName: string
  totalStakeSol: number
  totalLeftToCapSol: number
  marinadeStakeSol: number
  marinadeLeftToCapSol: number
  validators: AuctionValidator[]
}
