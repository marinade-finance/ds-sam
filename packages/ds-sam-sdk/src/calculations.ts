import { BidTooLowPenalty, AuctionValidator, Rewards, RevShare } from './types'

export const calcValidatorRevShare = (
  validator: { bidCpmpe: number | null, mevCommissionDec: number | null, inflationCommissionDec: number },
  rewards: Rewards
): RevShare => {
  const inflationPmpe = Math.max(0, rewards.inflationPmpe * (1 - validator.inflationCommissionDec))
  const mevPmpe = Math.max(0, rewards.mevPmpe * (1 - (validator.mevCommissionDec ?? 1)))
  const bidPmpe = Math.max(0, validator.bidCpmpe ?? 0)
  return {
    totalPmpe: inflationPmpe + mevPmpe + bidPmpe,
    inflationPmpe,
    mevPmpe,
    bidPmpe,
    auctionEffectiveBidPmpe: NaN,
    bidTooLowPenaltyPmpe: NaN,
    effParticipatingBidPmpe: NaN,
  }
}

export type BondRiskFeeConfig = {
  minBondEpochs: number
  idealBondEpochs: number
  minBondBalanceSol: number
  bondRiskFeeMult: number
}

export type BondRiskFeeResult = {
  bondForcedUndelegation: {
    base: number
    coef: number
    value: number
    feePmpe: number
  }
  bondRiskFee: number
  paidUndelegationSol: number
}

export const calcBondRiskFee = (
  cfg: BondRiskFeeConfig,
  validator: AuctionValidator,
): BondRiskFeeResult | null => {
  const { revShare } = validator
  const projectedActivatedStakeSol = validator.marinadeActivatedStakeSol - validator.values.paidUndelegationSol
  const minBondCoef = (revShare.totalPmpe + cfg.minBondEpochs * revShare.effParticipatingBidPmpe) / 1000
  const bondBalanceSol = validator.bondBalanceSol ?? 0
  if (bondBalanceSol < projectedActivatedStakeSol * minBondCoef) {
    const idealBondCoef = (revShare.totalPmpe + cfg.idealBondEpochs * revShare.effParticipatingBidPmpe) / 1000
    const feePmpe = revShare.inflationPmpe + revShare.mevPmpe + revShare.auctionEffectiveBidPmpe
    // always: base >= 0, since idealBondCoef >= minBondCoef, since idealBondEpochs >= minBondEpochs
    const base = projectedActivatedStakeSol - bondBalanceSol / idealBondCoef
    const coef = 1 - (feePmpe / 1000) / idealBondCoef
    let value = coef > 0 ? Math.min(projectedActivatedStakeSol, base / coef) : projectedActivatedStakeSol
    // always: value <= projectedActivatedStakeSol
    if (projectedActivatedStakeSol - value < cfg.minBondBalanceSol / (revShare.totalPmpe / 1000)) {
      value = projectedActivatedStakeSol
    }
    const bondRiskFee = cfg.bondRiskFeeMult * value * feePmpe / 1000
    const paidUndelegationSol = cfg.bondRiskFeeMult * value
    if (!isFinite(bondRiskFee)) {
      throw new Error(`bondRiskFee has to be finite`)
    }
    return {
      bondForcedUndelegation: { base, coef, value, feePmpe },
      bondRiskFee,
      paidUndelegationSol,
    }
  } else {
    return null
  }
}

export const calcEffParticipatingBidPmpe = (revShare: { inflationPmpe: number, mevPmpe: number }, winningTotalPmpe: number): number => {
  return Math.max(0, winningTotalPmpe - revShare.inflationPmpe - revShare.mevPmpe)
}

export type BidTooLowPenaltyResult = {
  bidTooLowPenalty: BidTooLowPenalty
  bidTooLowPenaltyPmpe: number
  paidUndelegationSol: number
}

export const calcBidTooLowPenalty = (
  bidTooLowPenaltyHistoryEpochs: number,
  winningTotalPmpe: number,
  validator: AuctionValidator
): BidTooLowPenaltyResult => {
  const { bidTooLowPenalty, revShare, auctions } = validator
  const historicalPmpe = auctions.slice(0, bidTooLowPenaltyHistoryEpochs).reduce(
    (acc, { effParticipatingBidPmpe }) => Math.min(acc, effParticipatingBidPmpe ?? Infinity),
    Infinity
  )
  const limit = Math.min(revShare.effParticipatingBidPmpe, historicalPmpe)
  const penaltyCoef = limit > 0 ? Math.min(1, Math.sqrt(1.5 * Math.max(0, (limit - revShare.bidPmpe) / limit))) : 0
  const bidTooLowPenaltyValue = {
    base: winningTotalPmpe + revShare.effParticipatingBidPmpe,
    coef: revShare.bidPmpe < 0.99999 * (auctions.map(({ bidPmpe }) => bidPmpe).find(x => x) ?? 0)
      ? penaltyCoef
      : 0
  }
  const bidTooLowPenaltyPmpe = bidTooLowPenaltyValue.coef * bidTooLowPenaltyValue.base
  const effPmpe = revShare.inflationPmpe + revShare.mevPmpe + revShare.auctionEffectiveBidPmpe
  const paidUndelegationSol = bidTooLowPenaltyPmpe * validator.marinadeActivatedStakeSol / effPmpe
  if (!isFinite(bidTooLowPenaltyPmpe)) {
    throw new Error('bidTooLowPenaltyPmpe has to be finite')
  }
  return {
    bidTooLowPenalty: bidTooLowPenaltyValue,
    bidTooLowPenaltyPmpe,
    paidUndelegationSol,
  }
}
