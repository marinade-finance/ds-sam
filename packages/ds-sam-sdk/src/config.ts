export enum InputsSource {
  APIS = 'APIS',
  FILES = 'FILES',
}

export type DsSamConfig = {
  inputsSource: InputsSource
  inputsCacheDirPath?: string
  cacheInputs?: boolean

  // TODO? split into nested config sections
  validatorsApiBaseUrl: string
  bondsApiBaseUrl: string
  tvlInfoApiBaseUrl: string
  blacklistApiBaseUrl: string
  snapshotsApiBaseUrl: string
  scoringApiBaseUrl: string
  overridesApiBaseUrl: string

  rewardsEpochsCount: number
  validatorsUptimeEpochsCount: number
  validatorsUptimeThresholdDec: number
  validatorsClientVersionSemverExpr: string
  validatorsMaxEffectiveCommissionDec: number
  bidTooLowPenaltyHistoryEpochs: number

  mndeDirectedStakeShareDec: number
  mndeStakeCapMultiplier: number
  maxMarinadeStakeConcentrationPerCountryDec: number
  maxMarinadeStakeConcentrationPerAsoDec: number
  maxNetworkStakeConcentrationPerCountryDec: number
  maxNetworkStakeConcentrationPerAsoDec: number
  maxMarinadeTvlSharePerValidatorDec: number
  spendRobustReputationMult: number | null
  spendRobustReputationDecayEpochs: number
  minSpendRobustReputation: number
  minScaledSpendRobustReputation: number
  maxSpendRobustReputation: number
  initialSpendRobustReputation: number
  minBondBalanceSol: number
  spendRobustReputationBondBoostCoef: number

  debugVoteAccounts: string[]
}

export const DEFAULT_CONFIG: DsSamConfig = {
  inputsSource: InputsSource.APIS,

  validatorsApiBaseUrl: 'https://validators-api.marinade.finance',
  bondsApiBaseUrl: 'https://validator-bonds-api.marinade.finance',
  tvlInfoApiBaseUrl: 'https://api.marinade.finance',
  blacklistApiBaseUrl: 'https://raw.githubusercontent.com/marinade-finance/delegation-strategy-2/master',
  overridesApiBaseUrl: 'https://raw.githubusercontent.com/marinade-finance/ds-sam-pipeline/main/overrides',
  snapshotsApiBaseUrl: 'https://snapshots-api.marinade.finance',
  scoringApiBaseUrl:  'https://scoring.marinade.finance',

  rewardsEpochsCount: 10,
  validatorsUptimeEpochsCount: 3,
  validatorsUptimeThresholdDec: 0.8,
  validatorsClientVersionSemverExpr: '>=1.18.15 || >=0.101.20013 <1.0.0',
  validatorsMaxEffectiveCommissionDec: 0.07,
  bidTooLowPenaltyHistoryEpochs: 3,

  mndeDirectedStakeShareDec: 0,
  mndeStakeCapMultiplier: 0.1,
  maxMarinadeStakeConcentrationPerCountryDec: 1,
  maxMarinadeStakeConcentrationPerAsoDec: 1,
  maxNetworkStakeConcentrationPerCountryDec: 0.3,
  maxNetworkStakeConcentrationPerAsoDec: 0.3,
  maxMarinadeTvlSharePerValidatorDec: 0.04,
  spendRobustReputationMult: null,
  spendRobustReputationDecayEpochs: 50,
  minSpendRobustReputation: -20,
  minScaledSpendRobustReputation: 40,
  maxSpendRobustReputation: 1000,
  initialSpendRobustReputation: 1,
  minBondBalanceSol: 0,
  spendRobustReputationBondBoostCoef: 0,

  debugVoteAccounts: [],
}
