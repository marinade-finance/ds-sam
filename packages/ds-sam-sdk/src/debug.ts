export class Debug {
  private infos: [string, string][] = []
  private events: string[] = []

  constructor (private readonly voteAccounts: Set<string>) { }

  getVoteAccounts () {
    return this.voteAccounts
  }

  pushInfo (context: string, info: string) {
    this.infos.push([context, info])
  }

  pushValidatorInfo (voteAccount: string, context: string, info: string) {
    if (this.voteAccounts.has(voteAccount)) {
      this.infos.push([context, `${voteAccount} ${info}`])
    }
  }

  pushEvent (event: string) {
    this.events.push(event)
  }

  pushValidatorEvent (voteAccount: string, event: string) {
    if (this.voteAccounts.has(voteAccount)) {
      this.events.push(`${voteAccount} ${event}`)
    }
  }

  pushValidatorSetEvent (voteAccounts: Set<string>, event: string) {
    this.voteAccounts.forEach(debugVoteAccount => {
      if (voteAccounts.has(debugVoteAccount)) {
        this.pushValidatorEvent(debugVoteAccount, event)
      }
    })
  }

  formatInfo (): string {
    return this.infos.map(([context, info]) => `DEBUG INFO - ${context}: ${JSON.stringify(info)}`).join('\n')
  }

  formatEvents (): string {
    return this.events.map(event => `DEBUG EVENT - ${event}`).join('\n')
  }
}
