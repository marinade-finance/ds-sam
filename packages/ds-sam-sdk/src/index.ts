export * from './config'
export * from './sdk'
export * from './types'
export * from './data-provider/data-provider.dto'
export * from './auction'
export * from './debug'
export * from './constraints'
