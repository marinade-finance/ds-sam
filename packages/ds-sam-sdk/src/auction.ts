import { AuctionData, AuctionResult, AuctionValidator } from './types'
import { AuctionConstraints, bondBalanceRequiredForCurrentStake } from './constraints'
import { Debug } from './debug'

const logValidators = (validators: AuctionValidator[]) => {
  console.log('validators -----------------------------')
  for (const validator of validators) {
    console.log(validator.voteAccount, validator.revShare.totalPmpe, validator.auctionStake.marinadeMndeTargetSol, validator.auctionStake.marinadeSamTargetSol)
  }
  console.log('----------------------------- validators')
}

export const EPSILON = 1e-4

export class Auction {

  constructor (private data: AuctionData, private constraints: AuctionConstraints, private debug: Debug) { }

  distributeMndeStake () {
    this.constraints.updateStateForMnde(this.data)
    this.debug.getVoteAccounts().forEach((voteAccount) => {
      const constraints = this.constraints.getValidatorConstraints(voteAccount)
      this.debug.pushValidatorEvent(voteAccount, `MNDE start constraints: ${constraints ? `${JSON.stringify(constraints.map(constraint => ({ ...constraint, validators: constraint.validators.length})))}` : 'NULL'}`)
    })

    let remEligibleValidators = this.data.validators.filter(validator => validator.mndeEligible)

    while (remEligibleValidators.length > 0) {
      const eligibleValidators = new Map(remEligibleValidators.map(validator => [validator.voteAccount, validator]))
      const eligibleVoteAccounts = new Set(eligibleValidators.keys())
      const { cap: evenDistributionCap } = this.constraints.getMinCapForEvenDistribution(eligibleVoteAccounts)
      console.log("MNDE distributing", evenDistributionCap, "to every validator in the group", eligibleValidators.size)

      for (const validator of eligibleValidators.values()) {
        validator.auctionStake.marinadeMndeTargetSol += evenDistributionCap
        this.data.stakeAmounts.marinadeRemainingMndeSol -= evenDistributionCap
      }

      // logValidators(remEligibleValidators)

      this.constraints.updateStateForMnde(this.data)
      remEligibleValidators = remEligibleValidators.filter((validator) => {
        const validatorCap = this.constraints.findCapForValidator(validator)
        if (validatorCap < EPSILON) {
          console.log('MNDE removing validator', validator.voteAccount, 'from the group because the cap has been reached')
          return false
        }
        return true
      })

      if (this.data.stakeAmounts.marinadeRemainingMndeSol < EPSILON) {
        console.log("MNDE No stake remaining to distribute")
        break
      } else {
        console.log("MNDE Stake remaining", this.data.stakeAmounts.marinadeRemainingMndeSol)
      }
    }
  }

  distributeSamStake () {
    this.constraints.updateStateForSam(this.data)
    this.debug.getVoteAccounts().forEach((voteAccount) => {
      const constraints = this.constraints.getValidatorConstraints(voteAccount)
      this.debug.pushValidatorEvent(voteAccount, `SAM start constraints: ${constraints ? `${JSON.stringify(constraints.map(constraint => ({ ...constraint, validators: constraint.validators.length})))}` : 'NULL'}`)
    })

    // logValidators(this.data.validators)

    let previousGroupPmpe = Infinity
    let group = null
    let winningTotalPmpe = Infinity
    let groups = 0
    let rounds = 0
    while (group = this.findNextPmpeGroup(previousGroupPmpe)) {
      groups++
      group.validators = group.validators.filter(validator => validator.samEligible)

      console.log('SAM ========= new round ==========')
      console.log('SAM', group.validators.length, 'validators eligible')
      this.debug.pushValidatorSetEvent(new Set(group.validators.map(({ voteAccount }) => voteAccount)), `assigned to PMPE group ${group.totalPmpe} with ${group.validators.length} eligible validators: ${group.validators.slice(0, 5).map(({ voteAccount }) => voteAccount).join(' ')}`)

      if (this.data.stakeAmounts.marinadeRemainingSamSol < EPSILON) {
        console.log("SAM No stake remaining to distribute")
        this.debug.pushEvent('SAM No stake remaining to distribute')
        break
      }

      while (group.validators.length > 0) {
        rounds++
        const remainingStakeToDistribute = this.data.stakeAmounts.marinadeRemainingSamSol
        const groupValidators = new Map(group.validators.map(validator => [validator.voteAccount, validator]))
        const groupVoteAccounts = new Set(groupValidators.keys())

        const { cap } = this.constraints.getMinCapForEvenDistribution(groupVoteAccounts)
        const evenDistributionCap = Math.min(cap, remainingStakeToDistribute / group.validators.length)
        console.log("SAM distributing", evenDistributionCap, "to every validator in the group", groupValidators.size)

        for (const validator of groupValidators.values()) {
          validator.auctionStake.marinadeSamTargetSol += evenDistributionCap
          this.data.stakeAmounts.marinadeRemainingSamSol -= evenDistributionCap
          winningTotalPmpe = validator.revShare.totalPmpe
          this.debug.pushValidatorEvent(validator.voteAccount, `received ${evenDistributionCap} SAM stake in PMPE group ${validator.revShare.totalPmpe} with ${groupValidators.size} validators`)
        }

        // logValidators(group.validators)

        this.constraints.updateStateForSam(this.data)
        group.validators = group.validators.filter((validator) => {
          const validatorCap = this.constraints.findCapForValidator(validator)
          if (validatorCap < EPSILON) {
            console.log('SAM removing validator', validator.voteAccount, 'from the group because the cap has been reached')
            return false
          }
          return true
        })

        if (this.data.stakeAmounts.marinadeRemainingSamSol < EPSILON) {
          console.log("SAM No stake remaining to distribute")
          this.debug.pushEvent('SAM No stake remaining to distribute')
          break
        } else {
          console.log("SAM Stake remaining", this.data.stakeAmounts.marinadeRemainingSamSol)
        }
      }

      previousGroupPmpe = group.totalPmpe
    }

    console.log("SAM rounds", rounds, "groups", groups)

    return winningTotalPmpe
  }

  setStakeUnstakePriorities () {
    this.data.validators.sort((a, b) => b.revShare.totalPmpe - a.revShare.totalPmpe)

    let currentGroupPmpe = NaN
    let currentStakePriority = 0
    this.data.validators.forEach(validator => {
      if (validator.revShare.totalPmpe === currentGroupPmpe) {
        validator.stakePriority = currentStakePriority
      } else {
        validator.stakePriority = ++currentStakePriority
        currentGroupPmpe = validator.revShare.totalPmpe
      }
    })

    this.data.validators
      .filter(({ mndeEligible, samEligible }) => !mndeEligible && !samEligible)
      .forEach(validator => validator.unstakePriority = 0)

    let bondsMaxIndex = 0
    this.data.validators
      .filter(({ unstakePriority }) => Number.isNaN(unstakePriority))
      .map(validator => ({
        validator,
        bondBalanceDiff: ((validator.bondBalanceSol ?? 0) - bondBalanceRequiredForCurrentStake(validator)) / validator.marinadeActivatedStakeSol
      }))
      .filter(({ bondBalanceDiff }) => bondBalanceDiff < 0) // Infinity and NaN filtered out too
      .sort((a, b) => a.bondBalanceDiff - b.bondBalanceDiff)
      .forEach(({ validator }, index) => bondsMaxIndex = validator.unstakePriority = index + 1)

    this.data.validators
      .filter(({ unstakePriority }) => Number.isNaN(unstakePriority))
      .map(validator => ({
        validator,
        stakeDiff: validator.marinadeActivatedStakeSol <= 0 ?
          1 :
          (validator.auctionStake.marinadeMndeTargetSol + validator.auctionStake.marinadeSamTargetSol - validator.marinadeActivatedStakeSol) / validator.marinadeActivatedStakeSol
      }))
      .sort((a, b) => a.stakeDiff - b.stakeDiff)
      .forEach(({ validator }, index) => validator.unstakePriority = bondsMaxIndex + index + 1)
  }

  setEffectiveBids(winningTotalPmpe: number) {
    this.data.validators.forEach(({ revShare }) => {
      if (revShare.totalPmpe < winningTotalPmpe) {
        revShare.auctionEffectiveBidPmpe = revShare.bidPmpe
      } else {
        revShare.auctionEffectiveBidPmpe = Math.max(0, winningTotalPmpe - revShare.inflationPmpe - revShare.mevPmpe)
      }
    })
  }

  evaluate (): AuctionResult {
    console.log('stake amounts before', this.data.stakeAmounts)
    this.debug.pushInfo('start amounts', JSON.stringify(this.data.stakeAmounts))
    this.debug.pushEvent('DISTRIBUTING MNDE STAKE')
    this.distributeMndeStake()

    this.debug.pushInfo('post MNDE amounts', JSON.stringify(this.data.stakeAmounts))
    console.log(`MNDE overflow: ${this.data.stakeAmounts.marinadeRemainingMndeSol}`)
    if (this.data.stakeAmounts.marinadeRemainingMndeSol > EPSILON) {
      this.debug.pushEvent(`MNDE overflow ${this.data.stakeAmounts.marinadeRemainingMndeSol} SOL will be distributed in SAM`)
      this.data.stakeAmounts.marinadeSamTvlSol += this.data.stakeAmounts.marinadeRemainingMndeSol
      this.data.stakeAmounts.marinadeRemainingSamSol += this.data.stakeAmounts.marinadeRemainingMndeSol
      this.data.stakeAmounts.marinadeMndeTvlSol -= this.data.stakeAmounts.marinadeRemainingMndeSol
      this.data.stakeAmounts.marinadeRemainingMndeSol = 0
    }

    this.debug.pushInfo('pre SAM amounts', JSON.stringify(this.data.stakeAmounts))
    this.debug.pushEvent('DISTRIBUTING SAM STAKE')
    const winningTotalPmpe = this.distributeSamStake()
    this.debug.pushEvent('STAKE DISTRIBUTED')

    console.log('stake amounts after', this.data.stakeAmounts)
    this.debug.pushInfo('end amounts', JSON.stringify(this.data.stakeAmounts))
    this.debug.pushInfo('winning total PMPE', winningTotalPmpe.toString())

    this.setStakeUnstakePriorities()
    this.setEffectiveBids(winningTotalPmpe)

    return {
      auctionData: this.data,
      winningTotalPmpe,
    }
  }

  findNextPmpeGroup (totalPmpe: number): { totalPmpe: number, validators: AuctionValidator[] } | null {
    console.log('finding next pmpe group...', totalPmpe)
    const nextGroupCandidates = this.data.validators.filter((validator) => validator.revShare.totalPmpe < totalPmpe)
    if (nextGroupCandidates.length === 0) {
      console.log('...no pmpe group remaining', totalPmpe)
      return null
    }
    const maxPmpe = nextGroupCandidates.reduce((max, validator) => Math.max(validator.revShare.totalPmpe, max), 0)
    const validators = nextGroupCandidates.filter((validator) => validator.revShare.totalPmpe === maxPmpe)

    console.log('...found next pmpe group', maxPmpe, validators.length)
    return {
      totalPmpe: maxPmpe,
      validators,
    }
  }
}
