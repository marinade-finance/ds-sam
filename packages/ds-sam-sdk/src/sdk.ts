import { DEFAULT_CONFIG, DsSamConfig, InputsSource } from './config'
import { DataProvider } from './data-provider/data-provider'
import {
  AggregatedData,
  AuctionValidator,
  AuctionData,
  ValidatorAuctionStake,
  AuctionResult,
  AuctionConstraintsConfig
} from './types'
import semver from 'semver'
import Decimal from 'decimal.js'
import { Auction } from './auction'
import { calcValidatorRevShare, ineligibleValidatorAggDefaults, validatorAggDefaults } from './utils'
import { AuctionConstraints } from './constraints'
import { Debug } from './debug'
import { SourceDataOverrides } from './data-provider/data-provider.dto'

export const defaultDataProviderBuilder = (config: DsSamConfig) => new DataProvider({ ...config }, config.inputsSource)

export class DsSamSDK {
  readonly config: DsSamConfig
  private readonly debug: Debug
  private readonly dataProvider: DataProvider

  constructor (config: Partial<DsSamConfig> = {}, dataProviderBuilder = defaultDataProviderBuilder) {
    this.config = { ...DEFAULT_CONFIG, ...config }
    this.dataProvider = dataProviderBuilder(this.config)
    this.debug = new Debug(new Set(this.config.debugVoteAccounts))
  }

  getAuctionConstraints ({ stakeAmounts }: AggregatedData, debug: Debug): AuctionConstraints {
    const { networkTotalSol, marinadeMndeTvlSol, marinadeSamTvlSol } = stakeAmounts
    const marinadeTotalTvlSol = marinadeMndeTvlSol + marinadeSamTvlSol
    const constraints: AuctionConstraintsConfig = {
      totalCountryStakeCapSol: networkTotalSol * this.config.maxNetworkStakeConcentrationPerCountryDec,
      totalAsoStakeCapSol: networkTotalSol * this.config.maxNetworkStakeConcentrationPerAsoDec,
      marinadeCountryStakeCapSol: marinadeTotalTvlSol * this.config.maxMarinadeStakeConcentrationPerCountryDec,
      marinadeAsoStakeCapSol: marinadeTotalTvlSol * this.config.maxMarinadeStakeConcentrationPerAsoDec,
      marinadeValidatorStakeCapSol: marinadeTotalTvlSol * this.config.maxMarinadeTvlSharePerValidatorDec,
    }
    this.debug.pushInfo('auction constraints', JSON.stringify(constraints))
    return new AuctionConstraints(constraints, debug)
  }

  transformValidators ({ validators, rewards, blacklist }: AggregatedData): AuctionValidator[] {
    let maxEpoch = 0
    const epochsTotals = validators.reduce((totals, { epochStats }) => {
      epochStats.forEach(({ epoch, totalActivatedStake, voteCredits }) => {
        const currentTotal = totals.get(epoch) ?? { weightedCredits: new Decimal(0), weight: new Decimal(0) }
        totals.set(epoch, {
          weightedCredits: currentTotal.weightedCredits.add(totalActivatedStake.mul(voteCredits)),
          weight: currentTotal.weight.add(totalActivatedStake)
        })
        maxEpoch = Math.max(maxEpoch, epoch)
      })
      return totals
    }, new Map<number, { weightedCredits: Decimal, weight: Decimal }>())

    const epochCreditsThresholds = new Map<number, number>()
    const minEpoch = maxEpoch - this.config.validatorsUptimeEpochsCount + 1
    for (let epoch = minEpoch; epoch <= maxEpoch; epoch++) {
      const epochTotals = epochsTotals.get(epoch)
      if (!epochTotals) {
        throw new Error(`Validator credits data for epoch ${epoch} not available`)
      }
      const threshold = epochTotals.weightedCredits.div(epochTotals.weight).mul(this.config.validatorsUptimeThresholdDec).toNumber()
      epochCreditsThresholds.set(epoch, threshold)
    }

    const minEffectiveRevSharePmpe = Math.max(0, rewards.inflationPmpe * (1 - this.config.validatorsMaxEffectiveCommissionDec))
    console.log('min rev share PMPE', minEffectiveRevSharePmpe)
    console.log('rewards', rewards)
    console.log('uptime thresholds', epochCreditsThresholds)
    this.debug.pushInfo('min effective rev share', minEffectiveRevSharePmpe.toString())
    this.debug.pushInfo('estimated rewards', JSON.stringify(rewards))

    return validators.map((validator): AuctionValidator => {
      const revShare = calcValidatorRevShare(validator, rewards)
      this.debug.pushValidatorInfo(validator.voteAccount, 'revenue share', JSON.stringify(revShare))
      const auctionStake: ValidatorAuctionStake = {
        externalActivatedSol: validator.totalActivatedStakeSol - validator.marinadeActivatedStakeSol,
        marinadeMndeTargetSol: 0,
        marinadeSamTargetSol: 0,
      }
      if (blacklist.has(validator.voteAccount)) {
        return { ...validator, revShare, auctionStake, ...ineligibleValidatorAggDefaults() }
      }
      if (!semver.satisfies(validator.clientVersion, this.config.validatorsClientVersionSemverExpr)) {
        return { ...validator, revShare, auctionStake, ...ineligibleValidatorAggDefaults() }
      }
      for (let epoch = minEpoch; epoch <= maxEpoch; epoch++) {
        const es = validator.epochStats.find(es => es.epoch === epoch)
        const threshold = epochCreditsThresholds.get(epoch)
        if (!es || !threshold || es.voteCredits < threshold) {
          return { ...validator, revShare, auctionStake, ...ineligibleValidatorAggDefaults() }
        }
      }
      if (validator.bondBalanceSol === null) {
        return { ...validator, revShare, auctionStake, ...ineligibleValidatorAggDefaults() }
      }
      const samEligible = revShare.totalPmpe >= minEffectiveRevSharePmpe
      const mndeEligible = revShare.inflationPmpe + revShare.mevPmpe >= minEffectiveRevSharePmpe

      return { ...validator, revShare, auctionStake, samEligible, mndeEligible, ...validatorAggDefaults() }
    })
  }

  async run (): Promise<AuctionResult> {
    const aggregatedData = await this.getAggregatedData()

    const constraints = this.getAuctionConstraints(aggregatedData, this.debug)
    const auctionData: AuctionData = { ...aggregatedData, validators: this.transformValidators(aggregatedData) }

    const auction = new Auction(auctionData, constraints, this.debug)
    const result = auction.evaluate()
    console.log(`==============================\n${this.debug.formatInfo()}\n${this.debug.formatEvents()}\n==============================`)
    return result
  }

  async getAggregatedData (dataOverrides: SourceDataOverrides | null = null): Promise<AggregatedData> {
    const sourceData = this.config.inputsSource === InputsSource.FILES ? this.dataProvider.parseCachedSourceData() : await this.dataProvider.fetchSourceData()
    return this.dataProvider.aggregateData(sourceData, dataOverrides)
  }
}
