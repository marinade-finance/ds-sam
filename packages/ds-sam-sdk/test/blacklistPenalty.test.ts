import { Auction } from '../src/auction'
import { Debug } from '../src/debug'

describe('setBlacklistPenalties', () => {
  it('applies penalties only to newly blacklisted validators', () => {
    const winningTotalPmpe = 1234.5

    const validators = [
      {
        voteAccount: 'newly-blacklisted',
        revShare: {
          effParticipatingBidPmpe: 100,
          blacklistPenaltyPmpe: 0
        },
        values: { samBlacklisted: true } as any,
        lastSamBlacklisted: false,
      },
      {
        voteAccount: 'still-blacklisted',
        revShare: {
          effParticipatingBidPmpe: 200,
          blacklistPenaltyPmpe: 0
        },
        values: { samBlacklisted: true } as any,
        lastSamBlacklisted: true,
      },
      {
        voteAccount: 'not-blacklisted',
        revShare: {
          effParticipatingBidPmpe: 300,
          blacklistPenaltyPmpe: 0
        },
        values: { samBlacklisted: false } as any,
        lastSamBlacklisted: false,
      },
      {
        voteAccount: 'newly-blacklisted-zero-eff',
        revShare: {
          effParticipatingBidPmpe: 0,
          blacklistPenaltyPmpe: 0
        },
        values: { samBlacklisted: true } as any,
        lastSamBlacklisted: false,
      },
    ] as any[]

    const data = { validators } as any

    const auction = new Auction(data, {} as any, {} as any, new Debug(new Set()))

    auction.setBlacklistPenalties(winningTotalPmpe)

    expect(
      validators.map(v => ({
        voteAccount: v.voteAccount,
        blacklistPenaltyPmpe: v.revShare.blacklistPenaltyPmpe
      }))
    ).toMatchInlineSnapshot(`
Array [
  Object {
    "voteAccount": "newly-blacklisted",
    "blacklistPenaltyPmpe": 1234.8,
  },
  Object {
    "voteAccount": "still-blacklisted",
    "blacklistPenaltyPmpe": 0,
  },
  Object {
    "voteAccount": "not-blacklisted",
    "blacklistPenaltyPmpe": 0,
  },
  Object {
    "voteAccount": "newly-blacklisted-zero-eff",
    "blacklistPenaltyPmpe": 1234.5,
  },
]
`)
  })
})
