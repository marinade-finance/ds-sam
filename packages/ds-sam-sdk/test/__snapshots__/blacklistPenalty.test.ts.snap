// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setBlacklistPenalties applies penalties only to newly blacklisted validators 1`] = `
[
  {
    "blacklistPenaltyPmpe": 1234.8,
    "voteAccount": "newly-blacklisted",
  },
  {
    "blacklistPenaltyPmpe": 0,
    "voteAccount": "still-blacklisted",
  },
  {
    "blacklistPenaltyPmpe": 0,
    "voteAccount": "not-blacklisted",
  },
  {
    "blacklistPenaltyPmpe": 1234.5,
    "voteAccount": "newly-blacklisted-zero-eff",
  },
]
`;
